drivers/adc/ad7689/ad7689.c

```diff
@@ -0,0 +1,291 @@
+	return &dev->configs[1];
+}
+
+static int32_t _ad7689_rac(struct ad7689_dev *dev,
+			   struct ad7689_config *config_in, struct ad7689_config *config_out,
+			   uint16_t *data)
+{
+	int32_t ret;
+	uint16_t cfg = 0;
+	uint8_t buf[4] = {0, 0, 0, 0};
+	uint16_t sz;
+	struct ad7689_config *c;
+
+	if (!dev)
+		return -EINVAL;
+
+	c = _ad7689_config_get(dev);
+
+	if (config_in) {
+		cfg |= field_prep(AD7689_CFG_CFG_MSK, 1);
+		cfg |= field_prep(AD7689_CFG_INCC_MSK, config_in->incc);
+		cfg |= field_prep(AD7689_CFG_INX_MSK, config_in->inx);
+		cfg |= field_prep(AD7689_CFG_BW_MSK, config_in->bw);
+		cfg |= field_prep(AD7689_CFG_REF_MSK, config_in->ref);
+		cfg |= field_prep(AD7689_CFG_SEQ_MSK, config_in->seq);
+		cfg |= field_prep(AD7689_CFG_RB_MSK, !config_in->rb);
+		cfg <<= 2;
+		buf[0] = cfg >> 8;
+		buf[1] = cfg;
+	}
+	sz = c->rb && config_out ? 4 : 2;
+	ret = spi_write_and_read(dev->spi_desc, buf, sz);
+	if (ret < 0)
+		return ret;
+
+	_ad7689_config_put(dev, config_in);
+
+	if (data) {
+		// by default, data width is 16-bits
+		*data = ((uint16_t)buf[0] << 8) | buf[1];
+
+		// handle 14-bit data width:
+		if (dev->id == ID_AD7949) {
+			// Bipolar samples are in two's complement, scale down to 14-bit
+			// by keeping the sign bit using division instead of implementation
+			// specific right shift. This works for unipolar samples too.
+			*data = (int16_t)*data / 4;
+		}
+	}
+
+	if (c->rb && config_out) {
+		if (dev->id == ID_AD7949)
+			cfg = (((uint32_t)buf[1] << 16) | ((uint32_t)buf[2] << 8) | buf[3]) >> 2;
+		else
+			cfg = ((uint16_t)buf[2] << 8) | buf[3];
+		cfg >>= 2;
+		config_out->incc = field_get(AD7689_CFG_INCC_MSK, cfg);
+		config_out->inx = field_get(AD7689_CFG_INX_MSK, cfg);
+		config_out->bw = field_get(AD7689_CFG_BW_MSK, cfg);
+		config_out->ref = field_get(AD7689_CFG_REF_MSK, cfg);
+		config_out->seq = field_get(AD7689_CFG_SEQ_MSK, cfg);
+		config_out->rb = !(bool)field_get(AD7689_CFG_RB_MSK, cfg);
+	}
+
+	return SUCCESS;
+}
+
+/***************************************************************************//**
+ * @brief Initialize the ad7689 driver and create a descriptor.
+ *
+ * @param dev - Device descriptor to create.
+ * @param init_param - Initialization parameters.
+ *
+ * @return Returns negative error code or SUCCESS in case of success.
+ *         Example: -EINVAL - Bad input parameters.
+ *                  -ENOMEM - Failed to allocate memory.
+ *                  SUCCESS - No errors encountered.
+*******************************************************************************/
+int32_t ad7689_init(struct ad7689_dev **dev,
+		    struct ad7689_init_param *init_param)
+{
+	struct ad7689_dev *d;
+	int32_t ret;
+
+	if (init_param->id > ID_AD7699)
+		return -EINVAL;
+
+	d = (struct ad7689_dev *)calloc(1, sizeof(*d));
+	if (!d)
+		return -ENOMEM;
+
+	d->id = init_param->id;
+	d->name = ad7689_device_name[d->id];
+
+	ret = spi_init(&d->spi_desc, &init_param->spi_init);
+	if (ret < 0)
+		goto error;
```

Comment:
is it required to call `spi_remove` in this case? or just `free(d)`?

Reply:
I added a patch to address this